### PR TITLE
Improve method resolution for numerics

### LINKED ISSUE #52: PEP 237 -- int/long Unification
Python 3 gets rid of the distinction between int and long, with int now behaving as long. One way would be to switch everything to BigInteger, but it might be more efficient to try and use Int32 as much as possible behind the scenes, and switch to BigInteger if we have to while keeping the visible type int at all times.

Probably worth doing some benchmarks to see if it's worthwhile first.

### PR DESCRIPTION
**Pull request by BCSharp**
This PR tackles some irregularities of method resolution for numerics, primarily the resolution involving `BigInteger` arguments.
There are three main issues addressed here.

The first one is actually rooted in the DLR. The DLR has a rule that it will prefer a parameter that is more specific, if such relationship between parameters exists. This is a good heuristic in general, for instance, given a parameter of type `object` and another of some class, the latter will be preferred. Or if one parameter is an interface and the other is a class implementing that interface, the latter will be preferred.

However, it is exactly the opposite in case of numerics. Given two numerical types, the wider (more generic) should be preferred, assuming a narrowing conversion is required. For instance, given `Int64` and `SByte` and argument of type `BigInteger`, `Int64` should be preferred. Or given `Int16` and `Byte` with argument of `Int32`, `Int16` should be preferred. But this is not how the DLR chooses. This augmented rule is so universal that I believe it deserves to be part of the DLR, but for now, I have implemented it at the IronPython level. After all, there is some legacy involved in the DLR so let's test things out in IronPython.

The second issue is the preference of numeric parameters when they do not have as strict generic-specific (or wider-narrower) relationship. Such situation happens when one is signed and the other unsigned. IronPython will then prefer the unsigned type. This is not by accident, such preference is explicitly codified in `Converter.PreferConvert` so clearly somebody thought this was a good idea at some point in the past. I could not find any comment or record justifying it, but this is how IronPython has been working for a long time so I am reluctant to change it. However, this causes unexpected behavior when the argument is a `BigInteger`. For instance, argument `Int32` choosing binding between `Int32` and `UInt32` will choose `Int32`, but `BigInteger` will choose `UInt32`, leading to overflow errors when the value is negative. Or, choosing binding between `Int64` and `UInt64`, argument `Int32` will choose `Int64`, but `BigInteger` will choose `UInt64`, with similar problems. Since now, after the int/long unification, `Int32` and `BigInteger` are treated the same, such differences in binding are problematic. Therefore I have broken the rule of preferring unsigned over signed if the argument is a `BigInteger` and the parameters are `Int32` or `Int64` (for narrower parameter types the old rule is still in effect). There are still some differences left between binding `Int32` and `BigInteger` but those are more corner cases, and in my opinion justified, given that `BigInteger` can really hold a very large value.

Finally, some strange behaviour was happening when the argument is a floating point, and the parameters are integer. #828 prohibited converting from floating point to `Int32`/`BigInteger`, but not to other integer types. So we had a situation that a floating point argument would prefer `Byte` over `Int32`, which is quite unexpected. Since I saw no reason to support automatic conversion from a floating point to any of the integer types and it makes the behaviour more regular, this is what I have implemented.

----

This is the last issue on my list for the int/long unification, so this PR closes #52.

### CODE REVIEW

**Comment by BCSharp** [comment]
See also https://github.com/IronLanguages/ironpython3/pull/891#issuecomment-647248928

**Review by slozier — APPROVED** [review]
Thanks! Looks good to me.

Unrelated comment, wish I could figure out those intermittent CI failures...

File: Src/IronPython/Runtime/Converter.cs [hunk 1]
```diff
@@ -732,35 +734,25 @@ public static partial class Converter {
...
-                //if (IsIntegral(fromType) && IsIntegral(toType)) return true;
 
-                //Check if there is an implicit convertor defined on fromType to toType
-                if (HasImplicitConversion(fromType, toType)) {
+            if (allowNarrowing >= PythonNarrowing.All) {
+                if (IsPythonNumeric(fromType) && IsPythonNumeric(toType)) {
+                    if (IsPythonFloatingPoint(fromType) && !IsPythonFloatingPoint(toType)) return false;
                     return true;
                 }
+                if (toType == Int32Type && HasPythonProtocol(fromType, "__int__")) return true;
```
[the remark this answers is not in the document]
Comment (BCSharp) [reply Src/IronPython/Runtime/Converter.cs:743]: I think you are right. I suppose `__index__` would be more appropriate here, but I do not have a strong strong preference in any specific way. I'd be OK too with simply removing it. Do you have any preference?

**Comment by BCSharp** [comment]
> wish I could figure out those intermittent CI failures...

They happen quite frequently locally at my ed too. It's always `test__ssl`. Maybe a flaky remote server. Maybe switching to another server would help.

File: Src/IronPython/Runtime/Converter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (slozier) [reply Src/IronPython/Runtime/Converter.cs:743]: I guess we could leave it as-is for now. Only becomes deprecated in 3.8 and removed in 3.10. Used this to test:
```py
class test:
   def __int__(self):
     return 1

chr(test())
```

**Pull request merged**